src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts

```diff
@@ -294,17 +313,20 @@ export module DotNet {
      *
      * @param identifier Identifies the globally-reachable function to invoke.
      * @param argsJson JSON representation of arguments to be passed to the function.
+     * @param byteArrays Byte array data extracted from the arguments for direct transfer.
      * @param resultType The type of result expected from the JS interop call.
      * @param targetInstanceId The instance ID of the target JS object.
-     * @returns JSON representation of the invocation result.
+     * @returns SerializedArgs containing the JSON representation of the invocation result and the associated byte arrays.
      */
-    invokeJSFromDotNet: (identifier: string, argsJson: string, resultType: JSCallResultType, targetInstanceId: number) => {
-      const returnValue = findJSFunction(identifier, targetInstanceId).apply(null, parseJsonWithRevivers(argsJson));
+    invokeJSFromDotNet: (identifier: string, argsJson: string, byteArrays: Uint8Array[], resultType: JSCallResultType, targetInstanceId: number): SerializedArgs => {
+      const returnValue = findJSFunction(identifier, targetInstanceId).apply(null, parseJsonWithRevivers(argsJson, byteArrays));
       const result = createJSCallResult(returnValue, resultType);
 
-      return result === null || result === undefined
+      const returnedByteArrays : Uint8Array[] = [];
+      var serializedJson = (result === null || result === undefined)
         ? null
-        : JSON.stringify(result, argReplacer);
+        : JSON.stringify(result, (key, value) => argReplacer(key, value, returnedByteArrays)); // TODO; confirm this works for blazor wasm
+      return new SerializedArgs(serializedJson, returnedByteArrays);
```

Comment:
Since this is single-threaded and doesn't allow developers to plug in custom serializers, there's an opportunity to make it a little bit more lightweight by not instantiating `returnedByteArrays` up front nor wrapping `argReplacer` in a further layer that runs for every nested object. Instead, the existing `argReplacer` function could take care of instantiating and populating some global `returnedByteArrays` object on every call. Example:

```ts
const serializedJson = (result === null || result === undefined)
  ? null
  : JSON.stringify(result, argReplacer);
const capturedByteArrays = globalCapturedByteArrays;
globalCapturedByteArrays = null;
return new SerializedArgs(serializedJson, capturedByteArrays);
```

... with:

```ts
// Pseudocode
function argReplacer() {
    // ... existing logic ...
    if (it is a byte array) {
        globalCapturedByteArrays = globalCapturedByteArrays || [];
        globalCapturedByteArrays[id] = newValue;
    }
}
```

The reason for suggesting this is trying to find ways to make the new functionality as close to zero cost as possible for existing apps that aren't using it.

Reply:
There's a couple of other places where this pattern would avoid the need for an extra layer of JSON revivers too.